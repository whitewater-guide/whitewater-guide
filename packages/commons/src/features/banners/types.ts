import { NamedNode, Node } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';
import { Connection } from '../types';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface Banner extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions: Node[];
  groups: Node[];
}

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
