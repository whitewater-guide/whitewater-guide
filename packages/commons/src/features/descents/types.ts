import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
