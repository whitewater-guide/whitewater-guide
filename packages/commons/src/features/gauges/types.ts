import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
