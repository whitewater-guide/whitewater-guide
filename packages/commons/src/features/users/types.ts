import { NamedNode, Timestamped } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';

export type Role = 'ADMIN' | 'EDITOR' | 'PREMIUM' | 'USER';

/**
 * This is graphql type
 */
export interface User extends NamedNode, Timestamped {
  avatar: string | null;
  email: string | null;
  admin: boolean;
  language: string;
  imperial: boolean;
  editorSettings: EditorSettings | null;

  purchasedRegions: Region[];
  purchasedGroups: Group[];
}

export interface EditorSettings {
  // @deprecated
  language: string;
}

export interface UserInput {
  name?: string;
  avatar?: string | null;
  language?: string;
  imperial?: boolean;
}
