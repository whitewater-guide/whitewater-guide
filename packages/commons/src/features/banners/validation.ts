import isNumber from 'lodash/isNumber';
import { customStruct } from '../../utils/validation';
import { BannerKind, BannerPlacement } from './types';

const struct = customStruct({
  ratio: (value: any) => {
    if (!isNumber(value)) {
      return 'Ratio must be a number';
    }
    if (value > 10 || value < 3) {
      return 'Ratio must be a number between 10 and 3';
    }
    return true;
  },
});

const BannerSourceStruct = struct.object({
  kind: struct.enum(Object.values(BannerKind)),
  ratio: 'ratio|null',
  src: 'nonEmptyString',
});

export const BannerInputStruct = struct.object({
  id: 'uuid|null',
  slug: 'slug',
  name: 'nonEmptyString',
  priority: 'integer',
  enabled: 'boolean',
  placement: struct.enum(Object.values(BannerPlacement)),
  source: BannerSourceStruct,
  link: 'https|null',
  extras: 'object|null',
  regions: ['node'],
  groups: ['node'],
});
