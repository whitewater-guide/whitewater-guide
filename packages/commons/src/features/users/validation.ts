// tslint:disable-next-line
import Joi from 'joi';
import { LANGUAGES } from '../../core';

export const UserInputSchema = Joi.object().keys({
  name: Joi.string().min(3).max(200).optional(),
  avatar: Joi.string().allow('').allow(null).optional(),
  language: Joi.any().valid(LANGUAGES).optional(),
  imperial: Joi.bool().optional(),
});
