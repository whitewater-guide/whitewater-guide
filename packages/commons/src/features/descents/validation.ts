import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
