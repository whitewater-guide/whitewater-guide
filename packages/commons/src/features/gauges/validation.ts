import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
