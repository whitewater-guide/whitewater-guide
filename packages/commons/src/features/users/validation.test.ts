import { createValidator } from '../../utils/validation';
import { UserInput } from './types';
import { UserInputStruct } from './validation';

const validator = createValidator(UserInputStruct);

type TestValue = [string, UserInput];

const correctValues: TestValue[] = [
  [
    'full value',
    { name: 'foo', avatar: 'bar', imperial: true, language: 'en' },
  ],
  ['empty avatar', { name: 'foo', avatar: '', imperial: true, language: 'en' }],
  [
    'null avatar',
    { name: 'foo', avatar: null, imperial: true, language: 'en' },
  ],
  ['no name', { avatar: 'bar', imperial: true, language: 'en' }],
  ['no avatar', { name: 'foo', imperial: true, language: 'en' }],
  ['no imperial', { name: 'foo', avatar: 'bar', language: 'en' }],
  ['no language', { name: 'foo', avatar: 'bar', imperial: true }],
];

const incorrectValues: TestValue[] = [
  ['empty name', { name: '' }],
  ['bad language', { language: 'xx' }],
  ['extra fields', { language: 'en', foo: 'bar' } as any],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  expect(validator(value)).not.toBeNull();
  expect(validator(value)).toMatchSnapshot();
});
