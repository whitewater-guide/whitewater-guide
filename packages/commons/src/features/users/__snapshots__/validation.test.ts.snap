// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad language 1`] = `
Object {
  "language": "Expected a value of type \`\\"en\\" | \\"ru\\" | \\"es\\" | \\"de\\" | \\"fr\\" | \\"pt\\" | \\"it\\" | undefined\` for \`language\` but received \`\\"xx\\"\`.",
}
`;

exports[`should be invalid for empty name 1`] = `
Object {
  "name": "Value must be non-empty string",
}
`;

exports[`should be invalid for extra fields 1`] = `
Object {
  "foo": "Expected a value of type \`undefined\` for \`foo\` but received \`\\"bar\\"\`.",
}
`;
